### PR TITLE
Redis (vapor 3) pub/sub support

### PR DESCRIPTION
**Pull request by Joannis**
Docs PR is in progress now

[docs](https://github.com/vapor/documentation/pull/232)

### CODE REVIEW

File: Sources/Redis/Pub+Sub.swift [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    
+    /// See `BaseStream.errorStream`
+    public var errorStream: ErrorHandler?
+    
+    /// Drains a Redis Client's parser of it's results
+    init(reading parser: DataParser) {
+        parser.drain { data in
+            guard
+                let array = data.array,
+                array.count == 3,
```
Comment (skyweb07) [inline Sources/Redis/Pub+Sub.swift]: what's this `3` magic number? could we name it?

Comment (Joannis) [reply Sources/Redis/Pub+Sub.swift]: [The wire protocol as described here](https://redis.io/topics/pubsub) has 3 fields.

- The type of notification
- The channel on which the notification is emitted
- The notification's payload

I'll describe this here 👍  But creating a global here isn't going to make things any more readable.
If `3` is renamed to `notificationPayloadSize` for example, we're left with the magic numbers `0`, `1` and `2` for array indexes.

**Comment by Joannis** [comment]
Thanks for the review by the way @skyweb07

File: Sources/Redis/Pub+Sub.swift (hunk 1 above)
Comment (twof) [reply Sources/Redis/Pub+Sub.swift]: Something like
```
enum WireProtocolFields: Int {
    case notificationType
    case notificationEmittedChannel
    case notificationPayload
}
```
maybe?

Lemme know if I'm bike shedding btw hahahaha

Comment (Joannis) [reply Sources/Redis/Pub+Sub.swift]: 😂 maybe. I'm not sure if this is ideal. But hey, if people think there mustn't be a magic number here I'm fine changing this.

File: Sources/Redis/Pub+Sub.swift [hunk 2]
```diff
@@ -0,0 +1,81 @@
+import Async
+
+extension RedisClient {
+    /// Subscribes to the given list of channels
+    public func subscribe(to channels: Set<String>) -> SubscriptionStream {
```
Comment (tanner0101) [inline Sources/Redis/Pub+Sub.swift]: we should have a convenience for subbing to one channel

File: Sources/Redis/Pub+Sub.swift [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            return RedisData(bulk: name)
+        }
+        
+        let command = RedisData.array(["SUBSCRIBE"] + channels)
+        
+        dataSerializer.inputStream(command)
+        
+        // Mark this client as being subscribed
+        // The client cannot be used for other commands now
+        self.subscribed = true
```
Comment (tanner0101) [inline Sources/Redis/Pub+Sub.swift]: should be `isSubscribed`

File: Sources/Redis/Pub+Sub.swift [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+                throw RedisError(.unexpectedResult(reply))
+            }
+            
+            return receivers
+        }
+    }
+}
+
+/// A message sent over pub/sub
+public struct ChannelMessage {
```
Comment (tanner0101) [inline Sources/Redis/Pub+Sub.swift]: separate files

File: Sources/Redis/Pub+Sub.swift [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    /// The channel this message was sent in
+    var channel: String
+    
+    /// The message itself
+    var message: RedisData
+}
+
+/// Outputs all notifications for a listening client's channels
+public final class SubscriptionStream: Async.OutputStream {
```
Comment (tanner0101) [inline Sources/Redis/Pub+Sub.swift]: separate file

**Pull request merged**